src/Thinktecture.Relay.Server/Transport/TenantConnectorAdapterRegistry.cs

```diff
@@ -0,0 +1,92 @@
+using System;
+using System.Collections.Concurrent;
+using System.Threading.Tasks;
+using Thinktecture.Relay.Server.Connector;
+using Thinktecture.Relay.Transport;
+
+namespace Thinktecture.Relay.Server.Transport
+{
+	internal class TenantConnectorAdapterRegistration<TRequest> : IDisposable
+		where TRequest : IRelayClientRequest
+	{
+		private readonly ITenantHandler<TRequest> _tenantHandler;
+		private readonly ITenantConnectorAdapter<TRequest> _tenantConnectorAdapter;
+
+		public Guid TenantId => _tenantConnectorAdapter.TenantId;
+
+		public TenantConnectorAdapterRegistration(ITenantHandler<TRequest> tenantHandler,
+			ITenantConnectorAdapter<TRequest> tenantConnectorAdapter)
+		{
+			_tenantHandler = tenantHandler ?? throw new ArgumentNullException(nameof(tenantHandler));
+			_tenantConnectorAdapter = tenantConnectorAdapter ?? throw new ArgumentNullException(nameof(tenantConnectorAdapter));
+
+			tenantHandler.RequestReceived += OnRequestReceived;
+		}
+
+		private async Task OnRequestReceived(object sender, TRequest @event) => await _tenantConnectorAdapter.RequestTargetAsync(@event);
+
+		public void Dispose()
+		{
+			_tenantHandler.RequestReceived -= OnRequestReceived;
+			(_tenantHandler as IDisposable)?.Dispose();
```

Comment:
Auch hier ggf. pattern matching?
`if (_tenantHandler is IDisposable d) d.Dispose();`